File: src/Plotly.NET/Traces/ObjectAbstractions/Box.fs
namespace Plotly.NET.TraceObjects

open Plotly.NET
open Plotly.NET.LayoutObjects
open DynamicObj
open System
open System.Runtime.InteropServices

/// Box type inherits from dynamic object (parent violin)
type Box() =
    inherit DynamicObj()

    /// Initialized Line object
    static member init
        (
            ?Visible: bool,
            ?Width: float,
            ?FillColor: Color,
            ?LineColor: Color,
            ?LineWidth: float
        ) =
        Box()
        |> Box.style (
            ?Visible = Visible,
            ?Width = Width,
            ?FillColor = FillColor,
            ?LineColor = LineColor,
            ?LineWidth = LineWidth
        )


    // Applies the styles to Line()
    static member style
        (
            ?Visible: bool,
            ?Width: float,
            ?FillColor: Color,
            ?LineColor: Color,
            ?LineWidth: float
        ) =
        fun (box: Box) ->

            let line =
                if LineColor.IsSome || LineWidth.IsSome then
                    Some(Line.init (?Color = LineColor, ?Width = LineWidth))
                else
                    None

            box
            |> DynObj.withOptionalProperty "visible" Visible
            |> DynObj.withOptionalProperty "width" Width
            |> DynObj.withOptionalProperty "fillcolor" FillColor
            |> DynObj.withOptionalProperty "line" line

File: src/Plotly.NET.CSharp/ObjectAbstractions/TraceObjects/Box.cs

using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Plotly.NET.CSharp;

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.Box Box
        (

            Optional<bool> Visible = default,
            Optional<double> Width = default,
            Optional<Color> FillColor = default,
            Optional<Color> LineColor = default,
            Optional<double> LineWidth = default
        ) => 
            Plotly.NET.TraceObjects.Box.init(

                Visible: Visible.ToOption(),
                Width: Width.ToOption(),
                FillColor: FillColor.ToOption(),
                LineColor: LineColor.ToOption(),
                LineWidth: LineWidth.ToOption()
            );
}

